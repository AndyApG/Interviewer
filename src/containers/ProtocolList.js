import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Swiper from 'react-id-swiper';

import ProtocolCard from '../components/Setup/ProtocolCard';
import { actionCreators as protocolActions } from '../ducks/modules/protocol';
import { actionCreators as sessionsActions } from '../ducks/modules/sessions';

/**
  * Display available protocols
  */
class ProtocolList extends Component {
  onClickNewProtocol = (protocol) => {
    if (!(protocol.type || protocol.path)) {
      console.log('protocol type and path not valid for:');
      console.log(protocol);
      return;
    }

    if (protocol.type === 'factory') {
      this.loadFactoryProtocol(protocol.path);
    } else {
      this.importRemoteProtocol(protocol.path);
    }
  }

  loadFactoryProtocol = (protocolPath) => {
    this.props.addSession();
    this.props.loadFactoryProtocol(protocolPath);
  }

  loadRemoteProtocol = (protocolPath) => {
    if (protocolPath) {
      this.props.addSession();
      this.props.loadProtocol(protocolPath);
    }
  }


  render() {
    const { protocols } = this.props;
    const params = {
      containerClass: 'protocol-list swiper-container',
      pagination: {
        el: '.swiper-pagination.protocol-list__pagination',
        type: 'bullets',
        clickable: true,
      },
      navigation: {
        nextEl: '.swiper-button-next',
        prevEl: '.swiper-button-prev',
      },
      loop: true,
    };

    return (
      <Swiper {...params}>
        { protocols.map(protocol => (
          <div key={protocol.path}>
            <ProtocolCard
              size="large"
              protocol={protocol}
              selectProtocol={() => this.onClickNewProtocol(protocol)}
            />
          </div>
        )) }
      </Swiper>
    );
  }
}

ProtocolList.propTypes = {
  addSession: PropTypes.func.isRequired,
  loadFactoryProtocol: PropTypes.func.isRequired,
  loadProtocol: PropTypes.func.isRequired,
  protocols: PropTypes.array.isRequired,
};

ProtocolList.defaultProps = {
};

function mapStateToProps(state) {
  return {
    protocols: state.protocols,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    addSession: bindActionCreators(sessionsActions.addSession, dispatch),
    loadProtocol: bindActionCreators(protocolActions.loadProtocol, dispatch),
    loadFactoryProtocol: bindActionCreators(protocolActions.loadFactoryProtocol, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(ProtocolList);
