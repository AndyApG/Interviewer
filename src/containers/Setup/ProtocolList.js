import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import Swiper from 'react-id-swiper';
import { size, map } from 'lodash';
import { NewSessionOverlay, ProtocolCard } from '../../components/Setup';
import { actionCreators as sessionActions } from '../../ducks/modules/sessions';
import { actionCreators as dialogActions } from '../../ducks/modules/dialogs';
import { actionCreators as importProtocolActions } from '../../ducks/modules/importProtocol';

/**
  * Display available protocols
  */
class ProtocolList extends Component {
  constructor(props) {
    super(props);
    this.state = {
      showNewSessionOverlay: false,
      selectedProtocol: null,
    };

    this.overlay = React.createRef();
  }
  onClickProtocolCard = (protocolUID) => {
    this.setState({
      showNewSessionOverlay: true,
      selectedProtocol: protocolUID,
    });
  }

  handleCreateSession = (caseId) => {
    this.props.addSession(caseId, this.state.selectedProtocol);
    this.handleCloseOverlay();

    // this.props.loadSession(protocol.path);
  }

  handleCloseOverlay = () => {
    this.setState({ showNewSessionOverlay: false, selectedProtocol: null });
  }

  render() {
    const {
      installedProtocols,
      // importProtocolStatus: {
      //   status,
      // },
    } = this.props;

    const params = {
      containerClass: 'protocol-list swiper-container',
      pagination: {
        el: '.swiper-pagination.protocol-list__pagination',
        type: 'bullets',
        clickable: true,
      },
      navigation: {
        nextEl: '.swiper-button-next.swiper-button-white',
        prevEl: '.swiper-button-prev.swiper-button-white',
      },
      on: {
        slideChange: this.handleSwipe,
      },
      loop: false,
      shouldSwiperUpdate: true,
      rebuildOnUpdate: true,
      slidesPerView: 'auto',
      centeredSlides: true,
    };

    return (
      <React.Fragment>
        { size(installedProtocols) > 0 ?
          <Swiper {...params} ref={(node) => { if (node) this.swiper = node.swiper; }}>
            { map(installedProtocols, (protocol, uid) => (
              <div key={uid}>
                <ProtocolCard
                  protocol={protocol}
                  selectProtocol={() => this.onClickProtocolCard(uid)}
                />
              </div>
            )) }
          </Swiper>
          :
          <div className="protocol-list protocol-list--empty">
            <div className="protocol-list--empty getting-started">
              <h1>No interview protocols installed</h1>
              <p>
                To get started, install an interview protocol on this device. To do this,
                click the button in the bottom left to pair with an instance of Server,
                import a protocol from a URL, or add a local .netcanvas file.
              </p>
            </div>
          </div>
        }
        <NewSessionOverlay
          handleSubmit={this.handleCreateSession}
          onClose={this.handleCloseOverlay}
          show={this.state.showNewSessionOverlay}
        />
      </React.Fragment>
    );
  }
}

ProtocolList.propTypes = {
  addSession: PropTypes.func.isRequired,
  loadSession: PropTypes.func.isRequired,
  installedProtocols: PropTypes.object.isRequired,
};

ProtocolList.defaultProps = {
};

function mapStateToProps(state) {
  return {
    installedProtocols: state.installedProtocols,
    importProtocolStatus: state.importProtocol,
  };
}

function mapDispatchToProps(dispatch) {
  return {
    addSession: bindActionCreators(sessionActions.addSession, dispatch),
    loadSession: bindActionCreators(sessionActions.loadSession, dispatch),
    openDialog: bindActionCreators(dialogActions.openDialog, dispatch),
    resetImportProtocol: bindActionCreators(importProtocolActions.resetImportProtocol, dispatch),
  };
}

export default connect(mapStateToProps, mapDispatchToProps)(ProtocolList);

export { ProtocolList as UnconnectedProtocolList };
