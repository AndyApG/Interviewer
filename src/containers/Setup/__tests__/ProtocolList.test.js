/* eslint-env jest */

import React from 'react';
import { shallow } from 'enzyme';
import { createStore } from 'redux';

import ProtocolList, { UnconnectedProtocolList } from '../ProtocolList';

const mockProps = {
  addSession: jest.fn(),
  loadFactoryProtocol: jest.fn(),
  loadProtocol: jest.fn(),
};

const mockProtocols = [{ name: 'Sample', path: 'sample.netcanvas', type: 'factory' }];

const mockStore = () =>
  createStore(() => ({
    protocols: mockProtocols,
  }));

describe('<ProtocolList />', () => {
  it('renders ok', () => {
    const component = shallow(<ProtocolList {...mockProps} store={mockStore()} />);

    expect(component).toMatchSnapshot();
  });

  describe('click handler', () => {
    let wrapper;

    beforeEach(() => {
      wrapper = shallow(<UnconnectedProtocolList {...mockProps} protocols={mockProtocols} />);
      mockProps.loadFactoryProtocol.mockClear();
      mockProps.loadProtocol.mockClear();
    });

    it('loads a remote protocol', () => {
      wrapper.instance().onClickNewProtocol(mockProtocols[0]);
      expect(mockProps.loadFactoryProtocol).toHaveBeenCalled();
    });

    it('loads a remote protocol', () => {
      const protocol = { ...mockProtocols[0], type: 'download', remoteId: '1' };
      wrapper.instance().onClickNewProtocol(protocol);
      expect(mockProps.loadFactoryProtocol).not.toHaveBeenCalled();
      expect(mockProps.loadProtocol).toHaveBeenCalledWith(protocol.path, null, protocol.remoteId);
    });
  });
});
