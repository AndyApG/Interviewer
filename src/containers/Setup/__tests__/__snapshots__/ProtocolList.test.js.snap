// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProtocolList /> renders ok 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Connect(ProtocolList)
    addSession={[MockFunction]}
    loadFactoryProtocol={[MockFunction]}
    loadProtocol={[MockFunction]}
    store={
      Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "addSession": [Function],
      "loadFactoryProtocol": [Function],
      "loadProtocol": [Function],
      "protocols": Array [
        Object {
          "isFactoryProtocol": true,
          "name": "Sample",
          "path": "sample.netcanvas",
        },
      ],
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "storeSubscription": Subscription {
        "listeners": Object {
          "clear": [Function],
          "get": [Function],
          "notify": [Function],
          "subscribe": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": [Function],
      },
    },
    "ref": null,
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "addSession": [Function],
        "loadFactoryProtocol": [Function],
        "loadProtocol": [Function],
        "protocols": Array [
          Object {
            "isFactoryProtocol": true,
            "name": "Sample",
            "path": "sample.netcanvas",
          },
        ],
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "storeSubscription": Subscription {
          "listeners": Object {
            "clear": [Function],
            "get": [Function],
            "notify": [Function],
            "subscribe": [Function],
          },
          "onStateChange": [Function],
          "parentSub": undefined,
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
          "unsubscribe": [Function],
        },
      },
      "ref": null,
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
